blocks/certifier.go

```diff
@@ -137,44 +131,39 @@ func NewCertifier(
 	}
 	c.collector = newCollector(c)
 
-	c.ctx, c.cancel = context.WithCancel(c.ctx)
 	return c
 }
 
 // Start starts the background goroutine for periodic pruning.
-func (c *Certifier) Start() {
+func (c *Certifier) Start(ctx context.Context) {
 	c.once.Do(func() {
+		ctx, c.stop = context.WithCancel(ctx)
 		c.eg.Go(func() error {
-			return c.run()
+			return c.run(ctx)
 		})
 	})
 }
 
 // Stop stops the outstanding goroutines.
 func (c *Certifier) Stop() {
-	c.cancel()
+	c.stopped.Store(true)
+	if c.stop == nil {
+		return // not started
+	}
```

Comment:
Otherwise calling `Stop` and then `Start` creates a service that cannot be stopped.

```suggestion
	if c.stop == nil {
		return // not started
	}
	c.stopped.Store(true)
```

Reply:
Yes, it's intended. Both the Generator and Certifier are single-use. They cannot be reused after being stopped. There's even a unit test that would fail after this change.

Reply:
Would it then make sense to call the `Stop` method `Close`? `Start`, `Stop` to me implies the ability to be able to restart it while `Close` means the object cannot be used afterwards any more.

Reply:
Perhaps :shrug:. It would be good to sit together and rethink all the services and plan to make them work and look similar.